include/features/feature_selector.h

```diff
@@ -0,0 +1,197 @@
+/**
+ * @file feature_selector.h
+ * @author Sean Massung
+ *
+ * All files in META are dual-licensed under the MIT and NCSA licenses. For more
+ * details, consult the file LICENSE.mit and LICENSE.ncsa in the root of the
+ * project.
+ */
+
+#ifndef META_FEATURE_SELECTOR_H_
+#define META_FEATURE_SELECTOR_H_
+
+#include <string>
+#include <vector>
+#include <memory>
+
+#include "util/disk_vector.h"
+#include "index/forward_index.h"
+
+namespace meta
+{
+namespace features
+{
+/**
+ * The base class that shows the feature selection interface for MeTA, allowing
+ * dimensionality reduction for documents as well as descriptions of classes by
+ * their useful features.
+ *
+ * Required config parameters:
+ * ~~~toml
+ * method = "corr-coef" # choose the feature selection algorithm
+ * prefix = "file-prefix"
+ * ~~~
+ *
+ * Optional config parameters:
+ * ~~~toml
+ * features-per-class = 20 # default
+ * ~~~
+ */
+class feature_selector
+{
+  public:
+    /**
+     * @param config
+     * @param idx
+     */
+    feature_selector(const std::string& prefix,
+                     std::shared_ptr<index::forward_index> idx);
+
+    /**
+     * Creates the state of this feature_selector if necessary; this logic is
+     * outside the constructor since it requires pure virtual functions
+     * implemented by deriving classes.
+     * @param features_per_class
+     */
+    void init(uint64_t features_per_class);
```

Comment:
Can't get it to compile for some reason with the templates, might have to ask you to look at it...